Adds a new style for function defs

pjrt opened the pull request:
This is a style that we are currently using and would like to keep using. It formats:

```scala
// maxColumn = 80
// this function is 98 chars long
def format_![T <: Tree](code: String, foo: Int)(f: A => B)(implicit ev: Parse[T], ev2: EC): String
```

to this:

```scala
def format_![T <: Tree](
    code: String,
    foo: Int
  )(f: A => B
  )(implicit ev: Parse[T],
    ev2: EC
  ): String
```

This format will only apply if the function would go over the maxColumn is single-lined. If it is under, it will be singled line.

pjrt commented on core/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala:
```diff
@@ -209,6 +223,7 @@ case class ScalafmtConfig(
     danglingParentheses: Boolean = false,
     poorMansTrailingCommasInConfigStyle: Boolean = false,
     bestEffortInDeeplyNestedCode: Boolean = false,
+    verticalMultiline: Boolean = false,
```
Name pending. It doesn't sound right but I can't think of anything else.

olafurpg replied:
Naming is hard. I would love to use an adt for styles supported in parameter lists with possible values bin pack, default and vertical multi-line. Currently it's a boolean in BinPack.

I'm OK with making this a boolean in scalafmt config, but the  name needs to be more specific. I suggest verticalMultilineAtDefinitionSite

pjrt replied:
It is odd because this is a whole style, but only for `def` sites (as opposed to a "style" like `IntelliJ`). This "mini-style" will also override things like continuation.defnSite (I'll make sure to mention it in the doc).

pjrt commented on core/src/main/scala/org/scalafmt/internal/Router.scala:
```diff
@@ -409,6 +409,52 @@ class Router(formatOps: FormatOps) {
         }
+
+      // Parameter opening for one parameter group. This format works
+      // on a group-by-group basis.
```
This doesn't sound right to me. I would like to have a holistic approach where I see the `def` as one thing, not as a bunch of `(`. I thought I could do so by using the `Decision` below, but for some reason a Decision for `(` (same as the top format case) just isn't picked up. 

Is there another way of doing this?

olafurpg replied:
Unfortunately no. The design is intentionally low level, for better or worse. Policies are the way to achieve what you want, which is to enforce behavior in upcoming format tokens. The first open paren of a def is where you enable the policy for the entire def/class parameter list.

olafurpg commented on core/src/main/scala/org/scalafmt/internal/Router.scala:
```diff
@@ -409,6 +409,52 @@ class Router(formatOps: FormatOps) {
+            oneLinePerArgPolicy.andThen(noNewLineOnParamSep)
+          // Find the very first leftParen in the `def`. This will tell us
+          // if we are in the first group or not.
+          val isFirstGroup =
```
I know this may feel hacky but it's totally idiomatic in the scalafmt code base.

olafurpg commented on core/src/main/scala/org/scalafmt/internal/Router.scala:
```diff
@@ -409,6 +409,52 @@ class Router(formatOps: FormatOps) {
+        val close = matchingParentheses(hash(open))
+        val indentParam = Num(4)
+        val indentSep = Num(2)
+        if (isTuple(leftOwner)) {
```
I would be very careful with the single line policy

pjrt replied:
This is actually not needed, left-over from when I was copy-and-pasting. Sorry for the noise.

pjrt commented on core/src/main/scala/org/scalafmt/internal/Router.scala:
```diff
@@ -409,6 +409,52 @@ class Router(formatOps: FormatOps) {
+            }.filter(_ == open).isDefined
+          val mod = if (isFirstGroup) Newline else NoSplit
+          Seq(
+            Split(mod, 0)
```
That would cause the split to happen partly. So it won't split on the first parameter group but it would on the second (if not splitting it causes it to go over the maxColumn). 

The test shows:

```scala
def format_![T <: Tree](code: String, foo: Int)(f: A => B
  )(implicit ev: Parse[T], ev2: EC): String
```

I need to somehow know, at the first `(` whether maxColumn has been violated or not and do one style or the other for the whole `def`.

olafurpg commented:
It's not possible to know the column width at this point. You only describe
the possible formatting options (including policies with expire tokens) and
their respective costs. The search will handle the max column.


On Wed, 18 Jan 2017 at 19:43, Pedro Rodriguez <notifications@github.com>
wrote:

> *@pjrt* commented on this pull request.
> ------------------------------
>
> In core/src/main/scala/org/scalafmt/internal/Router.scala
> <https://github.com/olafurpg/scalafmt/pull/682>:
>
> > +                  .withIndent(indentSep, close, Right)
> +                )
> +              )
> +          }
> +          val configStyleLike =
> +            oneLinePerArgPolicy.andThen(noNewLineOnParamSep)
> +          // Find the very first leftParen in the `def`. This will tell us
> +          // if we are in the first group or not.
> +          val isFirstGroup =
> +            leftOwner.tokens.find {
> +              case LeftParen() => true
> +              case _ => false
> +            }.filter(_ == open).isDefined
> +          val mod = if (isFirstGroup) Newline else NoSplit
> +          Seq(
> +            Split(mod, 0)
>
> That would cause the split to happen partly. So it won't split on the
> first parameter group but it would on the second (if not splitting it
> causes it to go over the maxColumn).
>
> The test shows:
>
> def format_![T <: Tree](code: String, foo: Int)(f: A => B
>
>
>   )(implicit ev: Parse[T], ev2: EC): String
>
> I need to somehow know, at the first ( whether maxColumn has been
> violated or not and do one style or the other for the whole def.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/olafurpg/scalafmt/pull/682>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABV8XXVm3Xxi5dlKI2dQ1cGxiJ92iVQ6ks5rTl1CgaJpZM4LnH3C>
> .
>

pjrt commented:
Ok I managed to get it to do what I wanted (it seems so obvious now). I've renamed the setting and added docs and comments all over the place. 

The interface for setting this is a bit odd, we should consider doing something about how binPack and now "vertical" work with one another.

olafurpg commented on core/src/test/resources/test/VericalMultiline.stat:
```diff
@@ -27,12 +27,15 @@ def format_![T <: Tree](code: String)(f: A => B)(implicit ev: Parse[T]): String
-  )(implicit ev: Parse[T]): String = 1
+    code: String,
+    code2: String
+  )(implicit ev: Parse[T], ex: D): String = 1
```
I would have expected there to be a new line before ex. Are you sure this is correct?

pjrt replied:
It is, this is the "before" case. The after case is below. These tests might be redundant though.

olafurpg replied:
Aha whoops! I didn't see that, on the phone -_-  good, then it makes sense!

olafurpg commented on core/src/main/scala/org/scalafmt/internal/Router.scala:
```diff
@@ -412,49 +413,58 @@ class Router(formatOps: FormatOps) {
+          Policy(oneLinePerArg.orElse(paramGroupSplitter), lastParen.end)
+
+        Seq(
+          Split(NoSplit, 0) // If it fits in one block, make it so
```
Split has an ignoreIf boolean field that you can set to true if this is not the first parameter list.

pjrt replied:
I removed that since I didn't need the whole "am I on the first block" thing. Now it is just "I am on a def, do either of these two policies: no split, single line or newline, vertical-style".

olafurpg replied:
Makes sense, this is clean!

olafurpg commented on core/src/main/scala/org/scalafmt/internal/Router.scala:
```diff
@@ -412,49 +413,58 @@ class Router(formatOps: FormatOps) {
-              )
+
+        @tailrec
+        def loop(token: Token): Option[Token] =
```
Can we move this to FormatOps with a clearer method name? Suggestion: lastClosingParam with return type Token.RightParen, no option since it can't be that by construction of the map. Exceptions are caught and dealt with higher up.

olafurpg replied:
Nevermind, this is good.

pjrt replied:
Oh, I was planning on moving it to FormatOps as you said tonight.

olafurpg commented:
Thank you for this contribution, I'll try have another release out this week :)
